File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/MecanumDriveBase.java
package org.firstinspires.ftc.teamcode.subsystems;

import com.qualcomm.robotcore.hardware.HardwareMap;
import com.qualcomm.robotcore.hardware.IMU;
import com.qualcomm.hardware.rev.RevHubOrientationOnRobot;
import org.firstinspires.ftc.robotcore.external.navigation.AngleUnit;
import com.qualcomm.robotcore.hardware.DcMotorEx;
import com.qualcomm.robotcore.hardware.DcMotorSimple.Direction;
import com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior;
import static com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior.*;


public class MecanumDriveBase {

    // -------------------------
    // Hardware
    
    
    private final DcMotorEx frontLeftMotor,
                            frontRightMotor,
                            backLeftMotor,
                            backRightMotor;

    // -------------------------
    // Configuration
    
    private static final String FRONT_LEFT_MOTOR_NAME  = "frontLeftDriveMotor";
    private static final String FRONT_RIGHT_MOTOR_NAME = "frontRightDriveMotor";
    private static final String BACK_LEFT_MOTOR_NAME   = "backLeftDriveMotor";
    private static final String BACK_RIGHT_MOTOR_NAME  = "backRightDriveMotor";

    private static final ZeroPowerBehavior FRONT_LEFT_MOTOR_ZERO_POWER_BEHAVIOR  = BRAKE;
    private static final ZeroPowerBehavior FRONT_RIGHT_MOTOR_ZERO_POWER_BEHAVIOR = BRAKE;
    private static final ZeroPowerBehavior BACK_LEFT_MOTOR_ZERO_POWER_BEHAVIOR   = BRAKE;
    private static final ZeroPowerBehavior BACK_RIGHT_MOTOR_ZERO_POWER_BEHAVIOR  = BRAKE;
    
    private static final Direction FRONT_LEFT_MOTOR_DIRECTION  = Direction.FORWARD;
    private static final Direction FRONT_RIGHT_MOTOR_DIRECTION = Direction.REVERSE;
    private static final Direction BACK_LEFT_MOTOR_DIRECTION   = Direction.FORWARD;
    private static final Direction BACK_RIGHT_MOTOR_DIRECTION  = Direction.REVERSE;

    public MecanumDriveBase(HardwareMap hardwareMap) {
        frontLeftMotor = hardwareMap.get(DcMotorEx.class, FRONT_LEFT_MOTOR_NAME);
        frontRightMotor = hardwareMap.get(DcMotorEx.class, FRONT_RIGHT_MOTOR_NAME);
        backLeftMotor = hardwareMap.get(DcMotorEx.class, BACK_LEFT_MOTOR_NAME);
        backRightMotor = hardwareMap.get(DcMotorEx.class, BACK_RIGHT_MOTOR_NAME);

        frontLeftMotor.setZeroPowerBehavior(FRONT_LEFT_MOTOR_ZERO_POWER_BEHAVIOR);
        frontRightMotor.setZeroPowerBehavior(FRONT_RIGHT_MOTOR_ZERO_POWER_BEHAVIOR);
        backLeftMotor.setZeroPowerBehavior(BACK_LEFT_MOTOR_ZERO_POWER_BEHAVIOR);
        backRightMotor.setZeroPowerBehavior(BACK_RIGHT_MOTOR_ZERO_POWER_BEHAVIOR);
        
        frontLeftMotor.setDirection(FRONT_LEFT_MOTOR_DIRECTION);
        frontRightMotor.setDirection(FRONT_RIGHT_MOTOR_DIRECTION);
        backLeftMotor.setDirection(BACK_LEFT_MOTOR_DIRECTION);
        backRightMotor.setDirection(BACK_RIGHT_MOTOR_DIRECTION);
    }
    
    public void driveFieldCentric(double drive, double strafe, double turn) {
        double thetaRadians = Math.atan2(drive, strafe);
        
        double power = Math.hypot(drive, strafe);

        double sinTheta = Math.sin(thetaRadians - Math.PI / 4.0);
        double cosTheta = Math.cos(thetaRadians - Math.PI / 4.0);
        double max = Math.max(Math.abs(sinTheta), Math.abs(cosTheta));

        double frontLeftPower  = power * cosTheta / max + turn;
        double frontRightPower = power * sinTheta / max - turn;
        double backLeftPower   = power * sinTheta / max + turn;
        double backRightPower  = power * cosTheta / max - turn;

        double turnMagnitude = Math.abs(turn);

        if (power + turnMagnitude > 1.0) {
            frontLeftPower  /= turnMagnitude;
            frontRightPower /= turnMagnitude;
            backLeftPower   /= turnMagnitude;
            backRightPower  /= turnMagnitude;
        }

        frontLeftMotor.setPower(frontLeftPower);
        frontRightMotor.setPower(frontRightPower);
        backLeftMotor.setPower(backLeftPower);
        backRightMotor.setPower(backRightPower);
    }
    
    public void driveRobotCentric(double drive, double strafe, double turn ) {
        double thetaRadians = Math.atan2(drive, strafe);
        
        double power = Math.hypot(drive, strafe);

        double sinTheta = Math.sin(thetaRadians - Math.PI / 4.0);
        double cosTheta = Math.cos(thetaRadians - Math.PI / 4.0);
        double max = Math.max(Math.abs(sinTheta), Math.abs(cosTheta));

        double frontLeftPower  = power * cosTheta / max + turn;
        double frontRightPower = power * sinTheta / max - turn;
        double backLeftPower   = power * sinTheta / max + turn;
        double backRightPower  = power * cosTheta / max - turn;

        double turnMagnitude = Math.abs(turn);

        if (power + turnMagnitude > 1.0) {
            frontLeftPower  /= turnMagnitude;
            frontRightPower /= turnMagnitude;
            backLeftPower   /= turnMagnitude;
            backRightPower  /= turnMagnitude;
        }

        frontLeftMotor.setPower(frontLeftPower);
        frontRightMotor.setPower(frontRightPower);
        backLeftMotor.setPower(backLeftPower);
        backRightMotor.setPower(backRightPower);
    }
}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/subsystems/MecanumDriveBase.kt
package org.firstinspires.ftc.teamcode.subsystems

import com.qualcomm.robotcore.hardware.DcMotor.ZeroPowerBehavior
import com.qualcomm.robotcore.hardware.DcMotorEx
import com.qualcomm.robotcore.hardware.DcMotorSimple
import com.qualcomm.robotcore.hardware.HardwareMap
import kotlin.math.*

class MecanumDriveBase(hardwareMap: HardwareMap) {
    private val frontLeftMotor  = hardwareMap[FRONT_LEFT_MOTOR_NAME]  as DcMotorEx
    private val frontRightMotor = hardwareMap[FRONT_RIGHT_MOTOR_NAME] as DcMotorEx
    private val backLeftMotor   = hardwareMap[BACK_LEFT_MOTOR_NAME]   as DcMotorEx
    private val backRightMotor  = hardwareMap[BACK_RIGHT_MOTOR_NAME]  as DcMotorEx

    init {
        frontLeftMotor.zeroPowerBehavior  = ZeroPowerBehavior.BRAKE
        frontRightMotor.zeroPowerBehavior = ZeroPowerBehavior.BRAKE
        backLeftMotor.zeroPowerBehavior   = ZeroPowerBehavior.BRAKE
        backRightMotor.zeroPowerBehavior  = ZeroPowerBehavior.BRAKE
    }

    var frontLeftPower: Double = 0.0
        private set

    var frontRightPower: Double = 0.0
        private set

    var backLeftPower: Double = 0.0
        private set

    var backRightPower: Double = 0.0
        private set

    fun setDrivePowers(drive: Double, strafe: Double, turn: Double) {
        val thetaRadians = atan2(drive, strafe)

        val power = hypot(drive, strafe)

        val sinTheta = sin(thetaRadians - Math.PI / 4.0)
        val cosTheta = cos(thetaRadians - Math.PI / 4.0)
        val max = max(abs(sinTheta), abs(cosTheta))

        frontLeftPower  = power * cosTheta / max + turn
        frontRightPower = power * sinTheta / max - turn
        backLeftPower   = power * sinTheta / max + turn
        backRightPower  = power * cosTheta / max - turn

        val turnMagnitude = abs(turn)

        if (power + turnMagnitude > 1.0) {
            frontLeftPower  /= turnMagnitude
            frontRightPower /= turnMagnitude
            backLeftPower   /= turnMagnitude
            backRightPower  /= turnMagnitude
        }
    }

    fun update() {
        frontLeftMotor.power  = frontLeftPower.coerceIn(-0.9, 0.9)
        frontRightMotor.power = frontRightPower.coerceIn(-0.9, 0.9)
        backLeftMotor.power   = backLeftPower.coerceIn(-0.9, 0.9)
        backRightMotor.power  = backRightPower.coerceIn(-0.9, 0.9)
    }

    companion object {
        // -------------------------
        // Configuration
        private const val FRONT_LEFT_MOTOR_NAME = "frontLeftDriveMotor"
        private const val FRONT_RIGHT_MOTOR_NAME = "frontRightDriveMotor"
        private const val BACK_LEFT_MOTOR_NAME = "backLeftDriveMotor"
        private const val BACK_RIGHT_MOTOR_NAME = "backRightDriveMotor"

        private val FRONT_LEFT_MOTOR_ZERO_POWER_BEHAVIOR = ZeroPowerBehavior.BRAKE
        private val FRONT_RIGHT_MOTOR_ZERO_POWER_BEHAVIOR = ZeroPowerBehavior.BRAKE
        private val BACK_LEFT_MOTOR_ZERO_POWER_BEHAVIOR = ZeroPowerBehavior.BRAKE
        private val BACK_RIGHT_MOTOR_ZERO_POWER_BEHAVIOR = ZeroPowerBehavior.BRAKE

        private val FRONT_LEFT_MOTOR_DIRECTION = DcMotorSimple.Direction.FORWARD
        private val FRONT_RIGHT_MOTOR_DIRECTION = DcMotorSimple.Direction.REVERSE
        private val BACK_LEFT_MOTOR_DIRECTION = DcMotorSimple.Direction.FORWARD
        private val BACK_RIGHT_MOTOR_DIRECTION = DcMotorSimple.Direction.REVERSE
    }
}
